[Frontend] [Core] Support for sharded tensorized models

tjohnson31415 opened the pull request:
Adds initial support for serializing and loading sharded vLLM models with Tensorizer. The implementation in this PR requires that the number of shards is equal to the number of target devices; there is one `.tensors` file for each shard. Therefore, the serialized model can only be loaded with the same sharding configuration that it was saved with.

Changes:
- to support multiple files, `TensorizerConfig.tensorizer_uri` can now contain a string template that will be formatted with the rank of the shard, eg. `.../model-rank-%03d.tensors`
- refactor `serialize_vllm_model` to take in encryption configuration (via the keyfile path) instead of generating a different key per shard
- updates `examples/tensorizer_vllm_model.py` to enable serializing a sharded model to multiple files
- plumbs through a `save_tensorized_model` function from `GPUExecutor` to  `TensorizerLoader` to enable distributed workers to run the serialization (analogous  to `save_sharded_state` used by `ShardedStateLoader`)
- Added `tensorize_vllm_model` utility function to the Tensorizer loader to easily produce a serialized vLLM model with encryption and/or sharding
- CUDA device index explicitly passed to `TensorDeserializer` so that tensors are created on the correct device
    - the internal thread pool seems to lose the tracking of the device index when the device is just `'cuda'`

FIX https://github.com/vllm-project/vllm/issues/4957

tjohnson31415 committed:
feat: support for sharded tensorized models

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
refactor: plumb through save_tensorized_model to remove monkey patch

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test/fix: update test_tensorizer

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

sangstar commented on examples/tensorize_vllm_model.py:
```diff
@@ -173,72 +171,94 @@ def deserialize():
+            **credentials)
+
+        engine = LLMEngine.from_engine_args(engine_args)
+        if args.tensor_parallel_size > 1:
+            engine.model_executor._run_workers("save_tensorized_model",
+                tensorizer_config = tensorizer_config,
+            )
+        else:
+            serialize_vllm_model(
+                engine.model_executor.driver_worker.model_runner.model,
+                tensorizer_config,
```
These seemingly represent the multi-GPU and single GPU serializations of the model respectively, although this is a bit unclear given the syntax. It may be worth considering adding some comments or wrapping them in to functions for clarity.

sangstar reviewed (commented):
Generally LGTM. I don't see any unit tests though -- am I missing something? Please make sure there are tests for the sharded serialization and loading. Thanks for the contribution!

tjohnson31415 committed:
test: add tensorizer tests with sharding

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
refactor: remove is_sharded from TensorizerConfig

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
refactor: write and use tensorizer_vllm_model util function

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
refactor: little bit of cleanup

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 commented:
Thanks for the review @sangstar!
You didn't miss anything, I hadn't included any new tests initially. I was hitting a problem where the process would hang during the test when attempting to create a second engine (i.e. one for serializing and then one for deserializing). I have added a test now though, using a subprocess to write out the serialized files as a workaround. I also did a bit of refactoring to support this with `tensorize_vllm_model` utility function which abstracts the overall procedure for creating a serialized model. This new function also includes some new comments on the differences between multi-GPU vs single-GPU. Please take another look at your convenience.

tjohnson31415 committed:
fix: add tensor parallel arg to deserialize()

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

sangstar commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -400,33 +410,65 @@ def is_vllm_tensorized(tensorizer_config: "TensorizerConfig") -> bool:
+            stream.write(encryption_params.key)
+
+    engine = LLMEngine.from_engine_args(engine_args)
+    if isinstance(engine.model_executor, DistributedGPUExecutor):
```
You use a couple of different ways to specify "this is using sharding" depending on the function. In this instance, your way of doing this is

```python
isinstance(engine.model_executor, DistributedGPUExecutor)
```

In `serialize_vllm_model()`, you use:

```python
re.search(r'%0\dd', output_file)
```

It may be a good idea to refactor this into an attribute for `TensorizerConfig` for readability. Did you have any misgivings about doing that? I see you removed `is_sharded`. It doesn't have to be a public attribute.

sangstar commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -240,7 +264,8 @@ def test_raise_value_error_on_invalid_load_format(vllm_runner):
 
 
-def test_tensorizer_with_tp(vllm_runner):
+@pytest.mark.skip("Requires multiple GPUs")
```
It would be better to use `@pytest.mark.skipif` in the case of the CI testing suite using (or ever will use) multiple GPUs. It feels hacky and incomplete for this test to skip unconditionally. Something like 

```python
@pytest.mark.skipif(not multiple_gpus_available())
```

would be better.

tjohnson31415 replied:
I copied this check from some of the lora tests that need multiple GPUs ([example here](https://github.com/vllm-project/vllm/blob/4238bc82f24d5887784b04a353ed93e2360623b4/tests/lora/test_llama.py#L83)). I agree that a `skipif` would be better; however, with the current bug around launching a second distributed engine during the test run, only one of the tests can pass at a time anyways. If we can find a fix for that, then I think we should use the `skipif` as you suggest.

sangstar commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -256,6 +281,65 @@ def test_tensorizer_with_tp(vllm_runner):
+        tensorizer_uri=model_path,
+        encryption_keyfile=key_path,
+    )
+    # FIXME: launching multiple multiprocessing servers within the same program
+    # results in a hang... launch serialization in a separate process as a work
+    # around
```
Will assume this is a WIP

tjohnson31415 replied:
I think the functionality supporting sharding is still useful even if the tests can't all pass together... Hopefully we can find a fix to this issue though. I'm planning to do a bit more digging into it to see if I can find anything, but let me know if you have ideas!

tjohnson31415 replied:
Found a work-around! The hang seems to be caused by the custom all reduce kernel not being completely cleaned up. The tests no longer hang if I use `disable_custom_all_reduce=True`. With that change I have removed this FIXME note and changed to skipif for the tests.

sangstar reviewed (changes requested):
Great work! Some requested changes. I'll give this some testing on my end soon and give you some feedback.

sangstar commented:
Can you also update `tensorize_vllm_model.py` in `examples/` with instructions for how to use your sharding feature? It's meant to be an unofficial vLLM-Tensorizer usage guide.

tjohnson31415 committed:
doc: add comment about sharding support in tensorize_vllm_model

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
refactor: add is_sharded back to TensorizerConfig

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: work around hang with disable_custom_all_reduce and use skipif

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

sangstar reviewed (commented):
Looking great! Just a few more nitpicks from me.

ywang96 commented:
@tjohnson31415 This is exciting! Have you done any performance eval on the model loading speed gain %?

cc @sroy745 if you have a chance to test this PR out

tjohnson31415 committed:
Apply suggestions from code review

Co-authored-by: Sanger Steel <sangersteel@gmail.com>

tjohnson31415 committed:
fmt/doc: minor changes

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
Merge remote-tracking branch 'upstream/main' into sharded-tensorizer

* upstream/main: (126 commits)
  [Bugfix][Frontend] Cleanup "fix chat logprobs" (#5026)
  [Bugfix] OpenAI entrypoint limits logprobs while ignoring server defined --max-logprobs (#5312)
  [Misc] Various simplifications and typing fixes (#5368)
  [ci] Fix Buildkite agent path (#5392)
  [Doc] Add documentation for FP8 W8A8 (#5388)
  Bump version to v0.5.0 (#5384)
  [Docs] Alphabetically sort sponsors (#5386)
  [Docs] Add Docs on Limitations of VLM Support (#5383)
  [ci] Mount buildkite agent on Docker container to upload benchmark results (#5330)
  [ci] Use small_cpu_queue for doc build (#5331)
  [Bugfix] Fix LLaVA-NeXT (#5380)
  [Feature][Frontend]:  Continued `stream_options` implementation also in CompletionRequest (#5319)
  [Model] Initial support for LLaVA-NeXT (#4199)
  [Misc] Improve error message when LoRA parsing fails (#5194)
  [misc][typo] fix typo (#5372)
  [Frontend][Misc] Enforce Pixel Values as Input Type for VLMs in API Server (#5374)
  [Misc] Update to comply with the new `compressed-tensors` config (#5350)
  [Bugfix] Fix KeyError: 1 When Using LoRA adapters (#5164)
  [Kernel][Misc] Use TORCH_LIBRARY instead of PYBIND11_MODULE for custom ops (#5047)
  [mis][ci/test] fix flaky test in test_sharded_state_loader.py (#5361)
  ...

tjohnson31415 commented:
@sangstar Thanks for all of the review comments! I believe I have addressed all of them. Let me know if anything else comes up from your testing, or if everything looks good now.

I rebased to resolve conflicts and now some tests are failing, but they seem to be unrelated (eg. jobs getting killed). The "Distributed Tests" are passing for me in my dev environment.

---

@ywang96. With regards to performance evaluation, I haven't done anything rigorous, but as a quick test:
I used the [ibm-granite/granite-34b-code-instruct](https://huggingface.co/ibm-granite/granite-34b-code-instruct) model sharded across 2 GPUs loaded from in-memory cached local files and looked at the server bootup times from the logs.
- Baseline without tensorizer with conversion from bfloat16 -> float16: ~30-32s
- Baseline without tensorizer and without conversion: ~21-22s
- With tensorizer loading from pre-converted float16: ~16-20s

So a nice speed up over converting the dtype of the tensors during the load with a smaller but still significant benefit even with the dtype conversion removed.

@sroy745 Let me know if you were able to test and have any feedback. Thanks.

sangstar commented:
> @sangstar Thanks for all of the review comments! I believe I have addressed all of them. Let me know if anything else comes up from your testing, or if everything looks good now.
> 
> I rebased to resolve conflicts and now some tests are failing, but they seem to be unrelated (eg. jobs getting killed). The "Distributed Tests" are passing for me in my dev environment.
> 
> @ywang96. With regards to performance evaluation, I haven't done anything rigorous, but as a quick test: I used the [ibm-granite/granite-34b-code-instruct](https://huggingface.co/ibm-granite/granite-34b-code-instruct) model sharded across 2 GPUs loaded from in-memory cached local files and looked at the server bootup times from the logs.
> 
> * Baseline without tensorizer with conversion from bfloat16 -> float16: ~30-32s
> * Baseline without tensorizer and without conversion: ~21-22s
> * With tensorizer loading from pre-converted float16: ~16-20s
> 
> So a nice speed up over converting the dtype of the tensors during the load with a smaller but still significant benefit even with the dtype conversion removed.
> 
> @sroy745 Let me know if you were able to test and have any feedback. Thanks.

Thanks for your awesome PR! LGTM. I can open up a new PR refactoring this once Tensorizer supports distributed serializing and deserializing natively, but this is a good use case of doing so with its current support. I'll defer to @ywang96 and @sroy745 for their reviews.

sangstar reviewed (approved):
Just tested it on my side. Great job! I'll defer to the other reviewers.

ywang96 reviewed (approved):
Overall LGTM! Thanks @sangstar for also reviewing this PR! Let's get this in for now.

One small ask to @sangstar (for later): When you do the refactoring, could you also add a short documentation page regarding using `tensorizer`? I know currently the documentation page refers it to the example script, but it'd be great to have some doc explaining the high-level idea and usage with the reference to the example script. Thanks!

sangstar commented:
> Overall LGTM! Thanks @sangstar for also reviewing this PR! Let's get this in for now.
> 
> One small ask to @sangstar (for later): When you do the refactoring, could you also add a short documentation page regarding using `tensorizer`? I know currently the documentation page refers it to the example script, but it'd be great to have some doc explaining the high-level idea and usage with the reference to the example script. Thanks!

Absolutely! That's on my and the team's radar. I plan on having a [Docs] PR for this out in the near future.

ywang96 committed:
Merge remote-tracking branch 'upstream/main' into sharded-tensorizer

sangstar commented:
Hey @ywang96 !!

After talking with my team, we're hesitant with doing something substantial with regard to documenting using `tensorizer` on vLLM's doc page, for a number of reasons, notably for maintenance and redundancy reasons (as the [example script](https://docs.vllm.ai/en/stable/getting_started/examples/tensorize_vllm_model.html) docstring summarizes things pretty well). I could add a brief overview (couple of sentences) and mostly point to `tensorizer`'s bespoke and maintained [documentation](https://github.com/coreweave/tensorizer) if that's desirable. Cheers!

ywang96 commented:
> Hey @ywang96 !!
> 
> After talking with my team, we're hesitant with doing something substantial with regard to documenting using `tensorizer` on vLLM's doc page for a number of reasons, notably for maintenance and redundancy concerns (as the [example script](https://docs.vllm.ai/en/stable/getting_started/examples/tensorize_vllm_model.html) docstring summarizes things pretty well). I could add a brief overview (couple of sentences) and mostly point to `tensorizer`'s bespoke and maintained [documentation](https://github.com/coreweave/tensorizer) if that's desirable. Cheers!

Thanks for keeping me posted about this - yea that works too!
